src/holochain/admin_conn.rs

```diff
@@ -1,51 +1,178 @@
+    AdminWebsocket, AppInfo, AuthorizeSigningCredentialsPayload, ConductorApiError,
+    SigningCredentials,
+};
 use holochain_conductor_api::{
-    AppAuthenticationTokenIssued, AppInterfaceInfo, IssueAppAuthenticationTokenPayload,
+    AppAuthenticationTokenIssued, AppInterfaceInfo, AppStatusFilter,
+    IssueAppAuthenticationTokenPayload,
 };
 use holochain_types::websocket::AllowedOrigins;
+use std::net::SocketAddr;
+use std::sync::Arc;
+use tokio::sync::RwLock;
+
+const ADMIN_WS_CONNECTION_MAX_RETRIES: usize = 1;
+
+/// A wrapper around AdminWebsocket that automatically handles reconnection
+/// when the connection is lost due to network issues or other failures.
+#[derive(Debug, Clone)]
+pub struct AdminConn {
+    /// The WebSocket URL to connect to
+    socket_addr: SocketAddr,
+
+    /// The handle to the AdminWebsocket connection - always contains a valid connection
+    handle: Arc<RwLock<Option<AdminWebsocket>>>,
+}
+
+impl AdminConn {
+    /// Creates a new [`AdminConn`] that will attempt to maintain an [`AdminWebsocket`] connection
+    /// to the specified socket address.
+    pub fn new(socket_addr: SocketAddr) -> Self {
+        Self {
+            socket_addr,
+            handle: Default::default(),
+        }
+    }
+
+    /// Allows calling a method on the [`AdminWebsocket`], with automatic reconnection if needed
+    pub async fn call<T>(
+        &self,
+        execute: impl Fn(AdminWebsocket) -> BoxFuture<'static, HcHttpGatewayResult<T>>,
+    ) -> HcHttpGatewayResult<T> {
+        for _ in 0..=ADMIN_WS_CONNECTION_MAX_RETRIES {
+            let admin_ws = self.get_admin_ws().await?;
+
+            match execute(admin_ws).await {
+                Ok(output) => return Ok(output),
+                Err(HcHttpGatewayError::HolochainError(ConductorApiError::WebsocketError(e))) => {
+                    tracing::warn!(
+                        ?e,
+                        "Detected admin websocket disconnection. Attempting to reconnect"
+                    );
+                    *self.handle.write().await = None;
+                    continue;
+                }
+                Err(e) => return Err(e),
+            }
+        }
+
+        Err(HcHttpGatewayError::UpstreamUnavailable)
+    }
 
-/// Placeholder for the admin connection.
-#[derive(Debug)]
-pub struct AdminConn;
+    async fn get_admin_ws(&self) -> HcHttpGatewayResult<AdminWebsocket> {
+        {
+            let lock = self.handle.read().await;
+
+            if let Some(admin_ws) = lock.as_ref() {
+                return Ok(admin_ws.clone());
+            }
+        }
+
+        let mut lock = self.handle.write().await;
+
+        // We might have been queued up behind another task that was holding the write lock, so we
+        // need to check again after obtaining the write lock. Reconnecting if another task has
+        // already reconnected risks closing the connection the other task just established.
+        if let Some(admin_ws) = lock.as_ref() {
```

Comment:
Only just noticed this. We need to check this again after obtaining a write lock. I've put the same change into the app websocket logic